scripts/issue/543/upgrade_ftm_emitting_vault.py

```diff
@@ -0,0 +1,92 @@
+from eth_abi import encode_abi
+from great_ape_safe import GreatApeSafe
+from helpers.addresses import registry
+from brownie import accounts, interface
+from rich.console import Console
+from brownie.test.managers.runner import RevertContextManager as reverts
+
+C = Console()
+
+NEW_LOGIC = registry.ftm.logic["theVaultWithoutTree"]
+VAULT_ADDRESS = registry.ftm.sett_vaults["bwFTM-WETH-wBTC"]
+DEV_PROXY = registry.ftm.badger_wallets.devProxyAdmin
+## Account with funds stuck due to integration
+USER = "0xB943cdb5622E7Bb26D3E462dB68Ee71D8868C940"
+
+def main(simulation="false"):
+    safe = GreatApeSafe(registry.ftm.badger_wallets.dev_multisig)
+
+    vault = interface.ITheVaultWithoutTree(VAULT_ADDRESS, owner=safe.account)
+ 
+    # record current vault information
+    prev_strategy = vault.strategy()
+    prev_guardian = vault.guardian()
+    prev_treasury = vault.treasury()
+
+    prev_badgerTree = vault.badgerTree()
+
+    prev_lifeTimeEarned = vault.lifeTimeEarned()
+    prev_lastHarvestedAt = vault.lastHarvestedAt()
+    prev_lastHarvestAmount = vault.lastHarvestAmount()
+    prev_assetsAtLastHarvest = vault.assetsAtLastHarvest()
+
+    prev_performanceFeeGovernance = vault.performanceFeeGovernance()
+    prev_performanceFeeStrategist = vault.performanceFeeStrategist()
+    prev_withdrawalFee = vault.withdrawalFee()
+    prev_managementFee = vault.managementFee()
+
+    prev_maxPerformanceFee = vault.maxPerformanceFee()
+    prev_maxWithdrawalFee = vault.maxWithdrawalFee()
+    prev_maxManagementFee = vault.maxManagementFee()
+
+    prev_toEarnBps = vault.toEarnBps()
+    prev_totalSupply = vault.totalSupply()
+    prev_getPricePerFullShare = vault.getPricePerFullShare()
```

Comment:
> i guess a storage hash of some sorts is not an option, given that the upgrade changes the storage no matter what?

Storage is not meant to change in any way, an upgrade that changes storage at the time of upgrade is going to cause problems

An upgrade that adds to storage or changes the meaning and functionality of storage is nuanced and has to be reviewed on a case by case basis

Reply:
@jorijnsmit, I tested this in a separate issue and it has a problem, when looping through the contract's attributes it calls them, including write functions. This means that all write functions will be executed or attempted to execute. It may not be the end of the world if we only run this under a simulation environment but it def introduces a lot of noise to the script with multiple reverts for unwanted calls.

There might be a way to skip write function calls but I haven't had a chance to look deeper into it. Contract.signatures gets them all and there's is no clear way to filter some of them out. Will keep thinking about it but, in the meantime, we are better off doing the manual checks imo.

Reply:
@jalbrekt85, you were right, that worked like a charm. Set functions or view functions with input parameters fail, I am printing them out in the `except` so that we can visually ensure that everything else is accounted.

In future implementations, we can add special cases for when variables change. Also, newly introduced variables should be tested separately or we can think of a more resilient approach for them. For the sake of time I pushed a very simple use of this that works.

Thanks!